src/debug_info.rs

```diff
@@ -1,30 +1,475 @@
+}
+
+impl<'ctx> DebugInfoBuilder<'ctx> {
+    pub(crate) fn new(module: &Module, allow_unresolved: bool) -> Self {
+        let builder = unsafe {
+            if allow_unresolved {
+                LLVMCreateDIBuilder(module.module.get())
+            } else {
+                LLVMCreateDIBuilderDisallowUnresolved(module.module.get())
+            }
+        };
+
+        DebugInfoBuilder {
+            builder,
+            _marker: PhantomData,
+        }
+    }
+
+    /// A DICompileUnit provides an anchor for all debugging information generated during this instance of compilation.
+    ///
+    /// * `language` - Source programming language
+    /// * `file` - File info
+    /// * `producer` - Identify the producer of debugging information and code. Usually this is a compiler version string.
+    /// * `is_optimized` - A boolean flag which indicates whether optimization is enabled or not.
+    /// * `flags` - This string lists command line options. This string is directly embedded in debug info output which may be used by a tool analyzing generated debugging information.
+    /// * `runtime_ver` - This indicates runtime version for languages like Objective-C.
+    /// * `split_name` - The name of the file that we'll split debug info out into.
+    /// * `kind` - The kind of debug information to generate.
+    /// * `dwo_id` - The DWOId if this is a split skeleton compile unit.
+    /// * `split_debug_inlining` - Whether to emit inline debug info.
+    /// * `debug_info_for_profiling` - Whether to emit extra debug info for profile collection.
+    pub fn create_compile_unit(
+        &self,
+        language: LLVMDWARFSourceLanguage,
+        file: DIFile<'ctx>,
+        producer: &str,
+        is_optimized: bool,
+        flags: &str,
+        runtime_ver: libc::c_uint,
+        split_name: &str,
+        kind: LLVMDWARFEmissionKind,
+        dwo_id: libc::c_uint,
+        split_debug_inlining: bool,
+        debug_info_for_profiling: bool,
+    ) -> DICompileUnit<'ctx> {
+        let producer = CString::new(producer).expect("Conversion to CString failed unexpectedly");
+        let flags = CString::new(flags).expect("Conversion to CString failed unexpectedly");
+        let split_name =
+            CString::new(split_name).expect("Conversion to CString failed unexpectedly");
+        let metadata_ref = unsafe {
+            LLVMDIBuilderCreateCompileUnit(
+                self.builder,
+                language,
+                file.metadata_ref,
+                producer.as_ptr(),
+                producer.as_bytes().len(),
+                if is_optimized { 0 } else { 1 },
+                flags.as_ptr(),
+                flags.as_bytes().len(),
+                runtime_ver,
+                split_name.as_ptr(),
+                split_name.as_bytes().len(),
+                kind,
+                dwo_id,
+                if split_debug_inlining { 0 } else { 1 },
+                if debug_info_for_profiling { 0 } else { 1 },
```

Comment:
Shouldn't these be 1 else 0?

Reply:
I got wrong values ( when llvm reported errors, and searching for "LLVMBool" got me [this](http://lists.llvm.org/pipermail/llvm-dev/2016-September/104766.html) and [this](https://ubiquitydotnet.github.io/Llvm.NET/articles/InternalDetails/marshal-LLVMBool.html) This seems incorrect now.

The problem was I was setting `is_definition` false on `create_function` and I got an unresolved `<temporary>` error: 

```
!4 = !DISubprogram(name: "main", linkageName: "main", scope: null, file: !1, line: 3, type: !5, scopeLine: 3, flags: DIFlagPublic, spFlags: 0, retainedNodes: !8)
!5 = !DISubroutineType(flags: DIFlagPublic, types: !6)
!6 = !{!7}
!7 = !DIBasicType(name: "()", size: 64, encoding: DW_ATE_signed, flags: DIFlagPublic)
!8 = <temporary!> !{}

and the error:
llc-8: output.ll:63:6: error: Expected '!' here
!8 = <temporary!> !{}
     ^
```

and setting it to true removed that `<temporary>`. I don't know what `retainedNodes` is, unfortunately.

Reply:
I got it working, I will try updating the Kaleidoscope example and update the PR.

Reply:
As I understand it, `LLVMBool` being `0` is success/true only applies to status codes and/or out params. So as an input function param, `1` should mean true. (I think)

It's madness that we even have to discuss this...

Good find on those links, btw.